src/Microsoft.ML.Data/Transforms/NAFilter.cs

```diff
@@ -204,25 +205,27 @@ private static bool TestType(ColumnType type)
             return null;
         }
 
-        protected override RowCursor GetRowCursorCore(Func<int, bool> predicate, Random rand = null)
+        protected override RowCursor GetRowCursorCore(IEnumerable<Schema.Column> colsNeeded, Random rand = null)
         {
-            Host.AssertValue(predicate, "predicate");
             Host.AssertValueOrNull(rand);
+            var predicate = RowCursorUtils.FromColumnsToPredicate(colsNeeded, OutputSchema);
 
-            bool[] active;
-            Func<int, bool> inputPred = GetActive(predicate, out active);
-            var input = Source.GetRowCursor(inputPred, rand);
+            Func<int, bool> inputPred = GetActive(predicate, out bool[] active);
+
+            var inputCols = Source.Schema.Where(x => inputPred(x.Index));
+            var input = Source.GetRowCursor(inputCols, rand);
             return new Cursor(this, input, active);
         }
 
-        public override RowCursor[] GetRowCursorSet(Func<int, bool> predicate, int n, Random rand = null)
+        public override RowCursor[] GetRowCursorSet(IEnumerable<Schema.Column> colsNeeded, int n, Random rand = null)
         {
-            Host.CheckValue(predicate, nameof(predicate));
             Host.CheckValueOrNull(rand);
+            var predicate = RowCursorUtils.FromColumnsToPredicate(colsNeeded, OutputSchema);
+
+            Func<int, bool> inputPred = GetActive(predicate, out bool[] active);
```

Comment:
So I see two reasons why you have this `FromColumnsToPredicate` method.
One is because in IRowToRowMapper we have 
`Func<int, bool> GetDependencies(Func<int, bool> predicate);` and `Row GetRow(Row input, Func<int, bool> active);`
And since you haven't change it, you need to create predicate.
I hope in later PRs you will change all predicates to list of columns, but I understand what you have to do it in chunks.

Second is case like this where you have `GetActive` function which accepts predicate and returns predicate. 
It's not part of any interface as far as I can see, and it's just a patter we have in our code.
I don't think such private local `GetActive` functions should continue to work with predicates.
If end goal to get rid of predicates in general, I don't see in what other PRs you will touch them again. Maybe when you decide to kill `FromColumnsToPredicate` method. Of maybe if you search for `Func<int,bool>` and notice their presence.

I'm fine with this code if you actually plan to remove all predicates you will eventually change `GetActive` method. Just make sure you take care of them.

Reply:
Thanks for writing this up, @ivanidzo4ka. I started cleaning up the predicated everywhere, and backed up because 1- the PR would be even bigger. 2- Testing it would be harder. 

Plan is to check this in, and start removing the predicates from GetDependencies, GetRow which is part of the public IRowToRowMapper. With those remove the utility function. 

---
In reply to: [247212899](https://github.com/dotnet/machinelearning/pull/1975#discussion_r247212899) [](ancestors = 247212899)